clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -329,7 +329,7 @@ private boolean maybeSendTransactionalRequest(long now) {
             return false;
 
         AbstractRequest.Builder<?> requestBuilder = nextRequestHandler.requestBuilder();
-        while (running) {
+        while (!forceClose) {
```

Comment:
The problem was basically that `initiateShutdown` set `running` to true, which prevented the transactional request from being sent. That in turn kept batched records stuck in the accumulator which prevented shutdown.

Reply:
nit (for documentation): `initiateClose` would set running to `false`.